crates/bevy_tasks/src/task_pool.rs

```diff
@@ -105,56 +131,82 @@ impl TaskPool {
-                let mut thread_builder = {
-                    let thread_name = if let Some(thread_name) = thread_name {
-                        format!("{} ({})", thread_name, i)
-                    } else {
-                        format!("TaskPool ({})", i)
-                    };
-                    thread::Builder::new().name(thread_name)
-                };
-                #[cfg(miri)]
-                let mut thread_builder = {
-                    let _ = i;
-                    let _ = thread_name;
-                    thread::Builder::new()
-                };
-
-                if let Some(stack_size) = stack_size {
-                    thread_builder = thread_builder.stack_size(stack_size);
-                }
-
-                thread_builder
-                    .spawn(move || {
-                        let shutdown_future = ex.run(shutdown_rx.recv());
-                        // Use unwrap_err because we expect a Closed error
-                        future::block_on(shutdown_future).unwrap_err();
-                    })
-                    .expect("Failed to spawn thread.")
-            })
-            .collect();
+        let compute_executor = Arc::new(async_executor::Executor::new());
+        let async_compute_executor = Arc::new(async_executor::Executor::new());
+        let io_executor = Arc::new(async_executor::Executor::new());
+
+        let compute_threads = builder.compute_threads.unwrap_or_else(num_cpus::get);
+        let io_threads = builder.io_threads.unwrap_or(0);
+        let async_compute_threads = builder.async_compute_threads.unwrap_or(0);
+
+        let mut threads = Vec::with_capacity(compute_threads + io_threads + async_compute_threads);
+        threads.extend((0..compute_threads).map(|i| {
+            let thread_builder = make_thread_builder(
+                builder.thread_name.as_deref(),
+                "Compute",
+                i,
+                builder.stack_size,
+            );
+            let compute = Arc::clone(&compute_executor);
+            let shutdown_rx = shutdown_rx.clone();
+
+            thread_builder
+                .spawn(move || {
+                    // Use unwrap_err because we expect a Closed error
+                    future::block_on(compute.run(shutdown_rx.recv())).unwrap_err();
+                })
+                .expect("Failed to spawn thread.")
+        }));
+        threads.extend((0..io_threads).map(|i| {
+            let thread_builder =
+                make_thread_builder(builder.thread_name.as_deref(), "IO", i, builder.stack_size);
+            let compute = Arc::clone(&compute_executor);
+            let io = Arc::clone(&io_executor);
+            let shutdown_rx = shutdown_rx.clone();
+
+            thread_builder
+                .spawn(move || {
+                    let future = io
+                        .run(shutdown_rx.recv())
+                        .or(compute.run(shutdown_rx.recv()))
+                        .or(shutdown_rx.recv());
```

Comment:
I should have mentioned on my last comment that the io executor and async_compute executor have two shutdown_rx's too.

Reply:
Done.